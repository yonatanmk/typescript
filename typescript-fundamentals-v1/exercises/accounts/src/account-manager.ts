interface User {
  email: string;
  password: string;
}

interface ConfirmedUser {
  email: string;
  password: string;
  isActive: boolean;
}

interface Admin {
  email: string;
  password: string;
  isActive: boolean;
  adminSince: Date;
}

export class AccountManager {
  users: User[] = new Array();

  /**
   * Create a new user account
   * @param email email address
   * @param password the user's password
   * @return the new user account. An admin must activate it using activateNewUser
   * @see this.activateNewUser
   */
  register(email: string, password: string): User {
    if(!email) throw 'Must provide an email';
    if(!password) throw 'Must provide a password';
    let user = { email, password };
    this.users.push(user);
    return user;
  }

  /**
   * Activate a new user account
   * @param approver The admin who's approving this new user
   * @param userToApprove Newly-registered user, who is to be activated
   * @return the updated user object, now activated
   */
  activateNewUser(approver: Admin, userToApprove: User): ConfirmedUser {
    if (!approver.adminSince) throw "Approver is not an admin!";
    let toConfirm = userToApprove as ConfirmedUser
    toConfirm.isActive = true;
    return toConfirm;
  }

  /**
   * Promote a normal user to admin
   * @param existingAdmin admin who is promoting another user
   * @param user an active user who you're making an admin
   * @return the updated user object, now can also be regarded as an admin
   */
  promoteToAdmin(existingAdmin: Admin, user: ConfirmedUser): Admin {
    if (!existingAdmin.adminSince) throw "Not an admin!";
    if (user.isActive !== true) throw "User must be active in order to be promoted to admin!";
    let newAdmin = user as Admin
    newAdmin.adminSince = new Date();
    return newAdmin;
  }
}
