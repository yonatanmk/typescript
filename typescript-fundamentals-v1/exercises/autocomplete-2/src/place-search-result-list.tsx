import * as React from 'react';


export const PlaceSearchResultList: React.SFC<any> = function() {
  return (
    <pre>{JSON.stringify(arguments)}</pre>
  )
}
