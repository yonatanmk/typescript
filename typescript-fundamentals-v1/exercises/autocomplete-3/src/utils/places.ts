export interface PlaceSummary {
  description: string;
  id: string;
  place_id: string;
}

export interface PlaceDetails {
  rating: number;
  icon: string;
  name: string;
  url: string;
  vicinity: string;
  website?: string;
}

/**
 * Search the google places API, for a particular term
 *
 * @export
 * @param {string} input Search term
 * @returns {Promise} promise that resolves to search results
 */
export function fetchPlaceSummaries(input: string): Promise<PlaceSummary[]> {
  return fetch(
    `http://localhost:3000/maps/api/place/autocomplete/json?types=establishment&input=${input}`
  )
    .then(response => response.json())
    .then(jsonData => {
      return jsonData.predictions as PlaceSummary[];
    });
}

export function fetchPlaceDetails(placeids: string[]): Promise<PlaceDetails[]> {
  return Promise.all(
    placeids.map(placeid => {
      return fetch(
        `http://localhost:3000/maps/api/place/details/json?placeid=${placeid}`
      )
        .then(response => response.json())
        .then(jsonData => jsonData.result as PlaceDetails);
    })
  );
}