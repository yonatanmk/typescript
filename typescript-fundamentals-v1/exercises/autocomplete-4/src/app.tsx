import * as React from 'react';
import { PlaceSearchContainer } from './place-search-container';

export const App: React.SFC = () => {
    return (
      <PlaceSearchContainer/>
    );
};
