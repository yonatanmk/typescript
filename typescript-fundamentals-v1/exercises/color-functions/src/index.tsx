import { hexToRgb, rgbToHex } from './color-utils';

export let color = {
  r: 255,
  g: 0,
  b: 0,
  get hex() : string {
    return rgbToHex(this.r, this.g, this.b);
  },
  set hex(hex: string) {
    let { r, g, b } = hexToRgb(hex);
    this.r = r;
    this.g = g;
    this.b = b;
  }
};