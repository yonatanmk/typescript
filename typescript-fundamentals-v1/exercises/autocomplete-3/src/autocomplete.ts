import { wait } from './utils/promise';
import { fetchPlaceSummaries, fetchPlaceDetails, PlaceDetails, PlaceSummary } from './utils/places';

export async function autocomplete(term: string): Promise<PlaceDetails[]> {
  console.log(`⏳ Beginning search for ${term}`);
  // Begin actual query API call
  let placeResults: PlaceSummary[] = await fetchPlaceSummaries(term);
  console.log(`✅ Completed search for ${term}`);
  let placeIds = placeResults.map(place => place.place_id);
  let places: PlaceDetails[] = await fetchPlaceDetails(placeIds);
  console.log(places);
  // Return the results (eventual value of the task)
  return places;
}
