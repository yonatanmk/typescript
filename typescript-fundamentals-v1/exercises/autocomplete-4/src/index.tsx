import * as React from 'react';
import { render } from 'react-dom';
// import { PlaceSearch } from './PlaceSearch';
import { App } from './app';
/**
 * Render the <PlaceSearch> component into the #root element
 */
const root = document.getElementById('root');
render(<App />, root);