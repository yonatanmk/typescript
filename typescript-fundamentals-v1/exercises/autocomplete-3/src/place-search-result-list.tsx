import * as React from 'react';
import { autocomplete } from './autocomplete';
import { shortUrl } from './utils/string';
import { PlaceDetails } from './utils/places';
import { PlaceSearchResult } from './place-search-result';

interface IPlaceSearchResultListProps {
  results: PlaceDetails[];
  inProgress: boolean;
  term: string;
  onSearchTermChanged?: (term: string) => void
}

export const PlaceSearchResultList: React.SFC<IPlaceSearchResultListProps> =
  ({ onSearchTermChanged, term, inProgress, results }) => {

    /**
     * There are a bunch of different things that we could end up
     * displaying as feedback to the user. Below we handle each of
     * four cases independently, taking care to always have searchResults
     * be an array.
     */
    let searchResults = null;
    if (term.length === 0) {
      // No search yet
      searchResults = [
        <li key='goahead'>Type something in above to search</li>
      ];
    } else if (inProgress) {
      // Search in progress
      searchResults = [
        <li key='searching' className="blue">Searching for {term}...</li>
      ];
    } else if (results.length === 0) {
      // No results found
      searchResults = [
        <li key='noresults' className="red">Sorry! No results for {term}.</li>
      ];
    } else {
      // Search complete
      let i = 0;
      searchResults = results.map(r => (
        <PlaceSearchResult key={r.url} {...r} />
      ));
    }
    return (
      <div>
        <h2>Search for a place</h2>
        <input type="search" placeholder="Search" onChange={e => onSearchTermChanged ? onSearchTermChanged(e.target.value) : () => {}} />
        <ul className="results">
          {searchResults}
        </ul>
      </div>
    );
  }
