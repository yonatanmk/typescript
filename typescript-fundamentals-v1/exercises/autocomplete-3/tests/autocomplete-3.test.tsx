import * as React from 'react';
import { PlaceSearchContainer } from '../src/place-search-container';
import * as renderer from 'react-test-renderer';
import { PlaceDetails } from '../src/utils/places';

let hasStudentSolution: boolean = true;

hasStudentSolution =
  JSON.stringify(renderer.create(<PlaceSearchContainer />).toJSON()) !==
  JSON.stringify({ type: 'p', props: {}, children: ['Replace this with a PlaceSearchResultList'] });

if (hasStudentSolution) {
  it('PlaceSearchContainer renders correctly', () => {
    const tree = renderer.create(<PlaceSearchContainer />).toJSON();
    expect(tree).toMatchSnapshot();
  });
} else {
  describe('Instructions', () => {
    test('Please implement the "autocomplete-3/src/place-search-container.tsx" component', () => {
      expect(true).toBeTruthy();
    });
  });
}
