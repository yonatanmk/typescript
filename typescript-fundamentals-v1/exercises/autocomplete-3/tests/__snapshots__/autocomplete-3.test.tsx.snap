// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceSearchContainer renders correctly 1`] = `
<div>
  <h2>
    Search for a place
  </h2>
  <input
    onChange={[Function]}
    placeholder="Search"
    type="search"
  />
  <ul
    className="results"
  >
    <li>
      Type something in above to search
    </li>
  </ul>
</div>
`;
